bindings/profilers/cpu.cc

```diff
@@ -30,26 +34,28 @@ CpuProfiler::CpuProfiler()
   // the pending sample and the vector to push symbolized samples to.
   async->data = static_cast<void*>(this);
   uv_sem_init(&sampler_thread_done, 1);
+
+  // Add cleanup hook to stop profiler upon Node process exit, otherwise
+  // SamplingThread could cause crashes by calling `Isolate::RequestInterrupt`
+  // during shutdown.
+  node::AddEnvironmentCleanupHook(isolate_, &cleanupProfiler, this);
 }
 
 CpuProfiler::~CpuProfiler() {
   uv_close(reinterpret_cast<uv_handle_t*>(async),
     [](uv_handle_t* handle){
       delete reinterpret_cast<uv_handle_t*>(handle);
     });
-  Stop();
-
-  uv_sem_wait(&sampler_thread_done);
+  StopAndWaitThread();
   uv_sem_destroy(&sampler_thread_done);
+
+  // Remove hook to avoid calling cleanup function on a destroyed object
+  node::RemoveEnvironmentCleanupHook(isolate_, &cleanupProfiler, this);
```

Comment:
I don't think we need this. The only time the destructor should be reached is through triggering the added cleanup hook, at which point the env is already in shutdown and will discard any remaining cleanup hooks automatically. Might actually even be problematic removing a hook while it's running given that the delete will be happening in the middle of the hook execution.

Reply:
I would like to merge this PR in the current state. The crash on shutdown occurs fairly often and this PR fixes it.
I could not determine what was causing the crash I saw with this PR, but it occurred only in an incorrect setup (`beforeExit` listener that got called infinitely).